breaking changes in gflags 2.2.0

Hey @theopolis 

Right now the osquery build fails with the latest gflags release (https://github.com/gflags/gflags/releases/tag/v2.2.0).

Full build log is here: https://gist.githubusercontent.com/ilovezfs/47de66dc430c5390abf3827ecc3e1a3b/raw/c09b71cdd3a06b13e86e98325f3aff32c33b8b37/gistfile1.txt

The error is
```
[  4%] Building CXX object third-party/gtest-1.7.0/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
cd /tmp/osquery-20161126-77522-l9y4dk/third-party/gtest-1.7.0 && /usr/local/Homebrew/Library/Homebrew/shims/super/clang++   -DBOOST_NETWORK_ENABLE_HTTPS -DBOOST_NO_CXX11_VARIADIC_TEMPLATES -DDARWIN -DDARWIN_10_12 -DHAS_SSL_TXT_TLSV1_2 -DNDEBUG -DOSQUERY_BUILD_DISTRO=10.12 -DOSQUERY_BUILD_PLATFORM=darwin -DOSQUERY_BUILD_SDK_VERSION=1.7.3 -DSTRIP_FLAG_HELP=1 -I/usr/local/opt/openssl/include -I/tmp/osquery-20161126-77522-l9y4dk/third-party/gtest-1.7.0/include -I/tmp/osquery-20161126-77522-l9y4dk/third-party/gtest-1.7.0  -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -mmacosx-version-min=10.12     -DGTEST_HAS_PTHREAD=1   -Wstrict-aliasing -Wno-missing-field-initializers -Wno-unused-local-typedef -Wno-unknown-warning-option -Wnon-virtual-dtor -Wchar-subscripts -Wpointer-arith -Woverloaded-virtual -Wformat -Wformat-security -Werror=format-security -Wabi-tag -fstack-protector-all -fPIE -fpie -fPIC -fpic -pipe -fdata-sections -ffunction-sections -mmacosx-version-min=10.12 -Os -march=x86-64 -mno-avx -o CMakeFiles/gtest_main.dir/src/gtest_main.cc.o -c /tmp/osquery-20161126-77522-l9y4dk/third-party/gtest-1.7.0/src/gtest_main.cc
/tmp/osquery-20161126-77522-l9y4dk/osquery/extensions/extensions.cpp:77:1: error: no matching constructor for initialization of 'google::FlagRegisterer'
EXTENSION_FLAG_ALIAS(socket, extensions_socket);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/osquery-20161126-77522-l9y4dk/include/osquery/flags.h:215:36: note: expanded from macro 'EXTENSION_FLAG_ALIAS'
#define EXTENSION_FLAG_ALIAS(a, n) OSQUERY_FLAG_ALIAS(std::string, a, n, 0, 1)
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/osquery-20161126-77522-l9y4dk/include/osquery/flags.h:208:43: note: expanded from macro 'OSQUERY_FLAG_ALIAS'
  static GFLAGS_NAMESPACE::FlagRegisterer oflag_##a(                  \
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:150:1: note: expanded from here
oflag_socket
^
/usr/local/include/gflags/gflags.h:439:3: note: candidate constructor template not viable: requires 5 arguments, but 6 were provided
  FlagRegisterer(const char* name,
  ^
/usr/local/include/gflags/gflags.h:432:23: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 6 were provided
class GFLAGS_DLL_DECL FlagRegisterer {
                      ^
/usr/local/include/gflags/gflags.h:432:23: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 6 were provided
/tmp/osquery-20161126-77522-l9y4dk/osquery/extensions/extensions.cpp:78:1: error: no matching constructor for initialization of 'google::FlagRegisterer'
EXTENSION_FLAG_ALIAS(timeout, extensions_timeout);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/osquery-20161126-77522-l9y4dk/include/osquery/flags.h:215:36: note: expanded from macro 'EXTENSION_FLAG_ALIAS'
#define EXTENSION_FLAG_ALIAS(a, n) OSQUERY_FLAG_ALIAS(std::string, a, n, 0, 1)
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/osquery-20161126-77522-l9y4dk/include/osquery/flags.h:208:43: note: expanded from macro 'OSQUERY_FLAG_ALIAS'
  static GFLAGS_NAMESPACE::FlagRegisterer oflag_##a(                  \
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<scratch space>:3:1: note: expanded from here
oflag_timeout
^
/usr/local/include/gflags/gflags.h:439:3: note: candidate constructor template not viable: requires 5 arguments, but 6 were provided
  FlagRegisterer(const char* name,
  ^
```

This is due to

https://github.com/gflags/gflags/pull/158
https://github.com/gflags/gflags/commit/46ea10f

I'm working around it with this patch:
```
diff --git a/include/osquery/flags.h b/include/osquery/flags.h
index 8112560..ca91255 100644
--- a/include/osquery/flags.h
+++ b/include/osquery/flags.h
@@ -206,7 +206,7 @@ class FlagAlias {
   FlagAlias<t> FLAGS_##a(#a, #t, #n, &FLAGS_##n);                     \
   namespace flags {                                                   \
   static GFLAGS_NAMESPACE::FlagRegisterer oflag_##a(                  \
-      #a, #t, #a, #a, &FLAGS_##n, &FLAGS_##n);                        \
+      #a, #a, #a, &FLAGS_##n, &FLAGS_##n);                        \
   const int flag_alias_##a = Flag::createAlias(#a, {#n, s, e, 0, 1}); \
   }
```

A complete fix would of course remove type from the osquery macro and its callers as well unless backwards compatibility with gflags 2.1.2 is needed.

--- a/include/osquery/flags.h
+++ b/include/osquery/flags.h
@@ -13,6 +13,7 @@
 #include <map>
 
 #include <boost/lexical_cast.hpp>
+#include <boost/noncopyable.hpp>
 
 #define GFLAGS_DLL_DEFINE_FLAG
 #define GFLAGS_DLL_DECLARE_FLAG
@@ -61,7 +62,7 @@ struct FlagInfo {
  * The osquery-specific gflags-like options define macro `FLAG` uses a Flag
  * instance to track the options data.
  */
-class Flag {
+class Flag : private boost::noncopyable {
  public:
   /*
    * @brief Create a new flag.
@@ -85,9 +86,6 @@ class Flag {
   /// Keep the ctor private, for accessing through `add` wrapper.
   Flag() {}
   virtual ~Flag() {}
-
-  Flag(Flag const&);
-  void operator=(Flag const&);
 
  public:
   /// The public flags instance, usable when parsing `--help`.
@@ -151,7 +149,10 @@ class Flag {
                          bool cli = false);
 
  private:
+  /// The container of all shell, CLI, and normal flags.
   std::map<std::string, FlagDetail> flags_;
+
+  /// A container for hidden or aliased (legacy, compatibility) flags.
   std::map<std::string, FlagDetail> aliases_;
 };
 
@@ -171,7 +172,9 @@ class FlagAlias {
     return *this;
   }
 
-  operator T() const { return boost::lexical_cast<T>(Flag::getValue(name_)); }
+  operator T() const {
+    return boost::lexical_cast<T>(Flag::getValue(name_));
+  }
 
   FlagAlias(const std::string& alias,
             const std::string& type,
@@ -180,6 +183,7 @@ class FlagAlias {
       : name_(name) {}
 
  private:
+  /// Friendly flag name.
   std::string name_;
 };
 }
@@ -187,31 +191,73 @@ class FlagAlias {
 /*
  * @brief Replace gflags' `DEFINE_type` macros to track osquery flags.
  *
- * @param type The `_type` symbol portion of the gflags define.
- * @param name The name symbol passed to gflags' `DEFINE_type`.
- * @param value The default value, use a C++ literal.
- * @param desc A string literal used for help display.
+ * Do not use this macro within the codebase directly! Use the subsequent macros
+ * that abstract the tail of boolean arguments into meaningful statements.
+ *
+ * @param type(t) The `_type` symbol portion of the gflags define.
+ * @param name(n) The name symbol passed to gflags' `DEFINE_type`.
+ * @param value(v) The default value, use a C++ literal.
+ * @param desc(d) A string literal used for help display.
+ * @param shell(s) Boolean, true if this is only supported in osqueryi.
+ * @param extension(e) Boolean, true if this is only supported in an extension.
+ * @param cli(c) Boolean, true if this can only be set on the CLI (or flagfile).
+ *   This helps documentation since flags can also be set within configuration
+ *   as "options".
+ * @param hidden(h) Boolean, true if this is hidden from help displays.
  */
-#define OSQUERY_FLAG(t, n, v, d, s, e, c, h)              \
-  DEFINE_##t(n, v, d);                                    \
-  namespace flags {                                       \
-  const int flag_##n = Flag::create(#n, {d, s, e, c, h}); \
+#define OSQUERY_FLAG(t, n, v, d, s, e, c, h)                                   \
+  DEFINE_##t(n, v, d);                                                         \
+  namespace flags {                                                            \
+  const int flag_##n = Flag::create(#n, {d, s, e, c, h});                      \
   }
 
+/*
+ * @brief Create a command line flag and configuration option.
+ *
+ * This is an abstraction around Google GFlags that allows osquery to organize
+ * the various types of "flags" used to turn features on and off and configure.
+ *
+ * A FLAG can be set within a `flagfile`, as a command line switch, or via
+ * a configuration's "options" key.
+ *
+ * @param t the type of flag, use the C++ symbol or literal type exactly.
+ * @param n the flag name as a symbol, write flagname instead of "flagname".
+ * @param v the default value.
+ * @param d the help description, please be concise.
+ */
 #define FLAG(t, n, v, d) OSQUERY_FLAG(t, n, v, d, 0, 0, 0, 0)
+
+/// See FLAG, but SHELL_FLAG%s are only available in osqueryi.
 #define SHELL_FLAG(t, n, v, d) OSQUERY_FLAG(t, n, v, d, 1, 0, 0, 0)
+
+/// See FLAG, but EXTENSION_FLAG%s are only available to extensions.
 #define EXTENSION_FLAG(t, n, v, d) OSQUERY_FLAG(t, n, v, d, 0, 1, 0, 0)
+
+/// See FLAG, but CLI_FLAG%s cannot be set within configuration "options".
 #define CLI_FLAG(t, n, v, d) OSQUERY_FLAG(t, n, v, d, 0, 0, 1, 0)
+
+/// See FLAG, but HIDDEN_FLAGS%s are not shown in --help.
 #define HIDDEN_FLAG(t, n, v, d) OSQUERY_FLAG(t, n, v, d, 0, 0, 0, 1)
 
-#define OSQUERY_FLAG_ALIAS(t, a, n, s, e)                             \
-  FlagAlias<t> FLAGS_##a(#a, #t, #n, &FLAGS_##n);                     \
-  namespace flags {                                                   \
-  static GFLAGS_NAMESPACE::FlagRegisterer oflag_##a(                  \
-      #a, #t, #a, #a, &FLAGS_##n, &FLAGS_##n);                        \
-  const int flag_alias_##a = Flag::createAlias(#a, {#n, s, e, 0, 1}); \
+/**
+ * @brief Create an alias to a command line flag.
+ *
+ * Like OSQUERY_FLAG, do not use this in the osquery codebase. Use the derived
+ * macros that abstract the tail of boolean arguments.
+ */
+#define OSQUERY_FLAG_ALIAS(t, a, n, s, e)                                      \
+  FlagAlias<t> FLAGS_##a(#a, #t, #n, &FLAGS_##n);                              \
+  namespace flags {                                                            \
+  static GFLAGS_NAMESPACE::FlagRegisterer oflag_##a(                           \
+      #a, #a, #a, &FLAGS_##n, &FLAGS_##n);                                     \
+  const int flag_alias_##a = Flag::createAlias(#a, {#n, s, e, 0, 1});          \
   }
 
+/// See FLAG, FLAG_ALIAS aliases a flag name to an existing FLAG.
 #define FLAG_ALIAS(t, a, n) OSQUERY_FLAG_ALIAS(t, a, n, 0, 0)
+
+/// See FLAG_ALIAS, SHELL_FLAG_ALIAS%es are only available in osqueryi.
 #define SHELL_FLAG_ALIAS(t, a, n) _OSQUERY_FLAG_ALIAS(t, a, n, 1, 0)
+
+/// See FLAG_ALIAS, EXTENSION_FLAG_ALIAS%es are only available to extensions.
 #define EXTENSION_FLAG_ALIAS(a, n) OSQUERY_FLAG_ALIAS(std::string, a, n, 0, 1)

--- a/tools/provision/formula/gflags.rb
+++ b/tools/provision/formula/gflags.rb
@@ -3,8 +3,8 @@ require File.expand_path("../Abstract/abstract-osquery-formula", __FILE__)
 class Gflags < AbstractOsqueryFormula
   desc "Library for processing command-line flags"
   homepage "https://gflags.github.io/gflags/"
-  url "https://github.com/gflags/gflags/archive/v2.1.2.tar.gz"
-  sha256 "d8331bd0f7367c8afd5fcb5f5e85e96868a00fd24b7276fa5fcee1e5575c2662"
+  url "https://github.com/gflags/gflags/archive/v2.2.0.tar.gz"
+  sha256 "466c36c6508a451734e4f4d76825cf9cd9b8716d2b70ef36479ae40f08271f88"
 
   bottle do
     root_url "https://osquery-packages.s3.amazonaws.com/bottles"

--- a/tools/provision/formula/glog.rb
+++ b/tools/provision/formula/glog.rb
@@ -5,6 +5,7 @@ class Glog < AbstractOsqueryFormula
   homepage "https://github.com/google/glog"
   url "https://github.com/google/glog/archive/v0.3.4.tar.gz"
   sha256 "ce99d58dce74458f7656a68935d7a0c048fa7b4626566a71b7f4e545920ceb10"
+  revision 1
 
   bottle do
     root_url "https://osquery-packages.s3.amazonaws.com/bottles"

